File: examples/python/order/handlers/create_order.py
"""CreateOrder command handler."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains
from .state import OrderState

from .exceptions import CommandRejectedError


def handle_create_order(command_book, command_any, state: OrderState, seq: int, log) -> angzarr.EventBook:
    if state.exists():
        raise CommandRejectedError("Order already exists")

    cmd = domains.CreateOrder()
    command_any.Unpack(cmd)

    if not cmd.customer_id:
        raise CommandRejectedError("Customer ID is required")
    if not cmd.items:
        raise CommandRejectedError("Order must have at least one item")

    subtotal = sum(item.quantity * item.unit_price_cents for item in cmd.items)

    log.info("creating_order", customer_id=cmd.customer_id, item_count=len(cmd.items))

    event = domains.OrderCreated(
        customer_id=cmd.customer_id,
        subtotal_cents=subtotal,
        created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )
    event.items.extend(cmd.items)

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )

File: examples/rust/order/src/handlers/create_order.rs
//! Handler for CreateOrder command.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{CreateOrder, OrderCreated, OrderState};
use common::{
    decode_command, make_event_book, now, require_not_empty, require_not_exists, require_positive,
    Result,
};
use prost::Message;

use crate::errmsg;

/// Handle the CreateOrder command.
pub fn handle_create_order(
    command_book: &CommandBook,
    command_data: &[u8],
    state: &OrderState,
    next_seq: u32,
) -> Result<EventBook> {
    require_not_exists(&state.customer_id, errmsg::ORDER_EXISTS)?;

    let cmd: CreateOrder = decode_command(command_data)?;

    require_not_empty(&cmd.items, errmsg::ITEMS_REQUIRED)?;
    for item in &cmd.items {
        require_positive(item.quantity, errmsg::QUANTITY_POSITIVE)?;
    }

    let subtotal: i32 = cmd
        .items
        .iter()
        .map(|i| i.quantity * i.unit_price_cents)
        .sum();

    let event = OrderCreated {
        customer_id: cmd.customer_id.clone(),
        items: cmd.items.clone(),
        subtotal_cents: subtotal,
        created_at: Some(now()),
        customer_root: cmd.customer_root.clone(),
        cart_root: cmd.cart_root.clone(),
    };

    let new_state = OrderState {
        customer_id: cmd.customer_id,
        items: cmd.items,
        subtotal_cents: subtotal,
        discount_cents: 0,
        loyalty_points_used: 0,
        payment_method: String::new(),
        payment_reference: String::new(),
        status: "pending".to_string(),
        customer_root: cmd.customer_root,
        cart_root: cmd.cart_root,
    };

    Ok(make_event_book(
        command_book.cover.clone(),
        next_seq,
        "type.examples/examples.OrderCreated",
        event.encode_to_vec(),
        "type.examples/examples.OrderState",
        new_state.encode_to_vec(),
    ))
}
